### DISCUSSION TITLE
How to pass values to PageContext via onBeforeRoute hook?

### DISCUSSION THREAD

**Discussion by thednp** [vikejs/vike#1306] [Discussion]
Thanks so much for this plugin.

I've read all the questions (especially [#256](https://github.com/vikejs/vike/issues/256#issue-1124987454)) and examples (especially the i18n-v1) and after I've tried everything, I give up, I will ask you guys.

I need to be able to pass in a token that I usually store in the *localStorage* from a previous page to the next page.

```js
function onBeforeRoute(pageContext) {
  let token;
 
  if (typeof process !== 'undefined') {
    token = pageContext?.token // server thing
  } else {
    token = localStorage.getItem('session_token');  // client thing
  }
  console.log(token);
  return {
    pageContext: {
      // similar to how you deal with locale, here's my token
      token
    }
  }
}
```

When I hover a `<Link>` the browser console shows the correct value, but on the VSCode console always gets unidentified.

Now I know this isn't the cleanest way, certainly far from a real solution, but I'm quite restricted to not using a server with express in it (my boss won't like me using a "monorepo like thing") like your auth-v1 example (which I also tried and it didn't work because my renderer/onBeforeRender wasn't correctly setup for that).

Also something like a globalContext would greatly help, probably, but you guys know best, so please point me to the right track.

Thank you for any suggestions!

**Comment by brillout** [comment]
Hm, if it's in localStorage then it's globally available and you should always be able to access it as long as you're on the client-side. So I'm not sure I see the problem.

That said, minimal reproduction welcome.

> I've read all the questions (especially [#256](https://github.com/vikejs/vike/issues/256#issue-1124987454)) and examples (especially the i18n-v1) and after I've tried everything, I give up, I will ask you guys.

👌

**Comment by thednp** [comment]
Thank you for taking the time to answer!

Yes, the value is globally available, however I'm trying to remove the jumpy ride in the UI from a pre-rendered stage of, for instance the User Account template with the Login form to the User Profile template, and I don't know how to modify this behavior.

**Reply by brillout** [reply]
Hm, I don't know. Seems very user-land-ish and not particularly related to Vike?

**Comment by thednp** [comment]
The best I can describe is: how can we do a pageContext with `{ token, user }` on first render (`onRenderHtml()`) without using a *server* folder with `express.js` inside?

Is it possible to send the token from client to server, from previously rendered page to a new page?

I hope this is clear.

**Reply by brillout** [reply]
You cannot access it at pre-render time so use a client side hook instead.

**Comment by thednp** [comment]
Thank you. I ended up implementing an express server.
